### DISCUSSION TITLE
Correlate output with specific test run using a plugin?

### DISCUSSION THREAD

**Discussion by eleanorjboyd** [pytest-dev/pytest#11672] [Q&A]
Hello again!

I had a quick question regarding output from tests, for context I work on python testing support in vscode. I have gotten asks on the vscode-python extension like [here](https://github.com/microsoft/vscode-python/issues/21781) to correlate output to a given test so I can display this output in the error pop up about the test. By output I mean print statements that are called during the run of this test and any logging based on the level the user defines.

I looked through the repo and tried a few things out with the plugin I use to collect pytest information but am currently seeing no way to connect output to a given test run when pytest is called to run a group of tests. For example, if the user selects to run 4 tests, and so I call pytest run once on all 4 of those, I wasn't seeing a clear way to separate output between each test with 100% confidence.

Is this correct or is there a way to correlate output to a test when using a pytest plugin? I am also open to ideas if you think of another way this could potentially be achieved.

Thank you as always!

**Comment by yashdeored** [comment]
When working with pytest and trying to associate output with specific tests in a tool or extension like the vscode-python extension, managing the output from tests can be challenging. Pytest, by default, aggregates the output of all tests executed, making it difficult to correlate individual output to specific tests when running multiple tests together.

The snippet you've shared demonstrates logging within a test. Setting the logging level to "INFO" (log_cli_level="INFO") will display INFO level logs during the test run when pytest is executed from the command line. However, this doesn't inherently associate this specific output with the test itself in a way that can be easily captured by an external tool like the vscode-python extension.

To correlate output with specific tests in a pytest plugin or extension, you might explore the following approaches:

1. Custom Output Handling: Develop a mechanism to intercept and manage output during test execution. This could involve intercepting print statements, capturing log messages, and associating them with the specific test that generated them. You might need to modify the pytest execution flow or capture the output stream for individual tests.

2. Parsing Test Execution Output: Analyze pytest's output during test execution. While it doesn't provide a direct mapping of output to specific tests, you might parse the output stream to identify patterns or markers that indicate the origin of the output (e.g., test names, identifiers).

3. Test Execution Isolation: Consider running tests individually or in a more isolated environment. This could potentially help in associating output with specific tests as there's less interference from other test executions.

4. Investigate pytest Plugins: Explore existing pytest plugins or develop a custom plugin that might offer functionalities to better manage or associate test outputs. Some plugins might provide hooks or mechanisms for capturing output per test.

5. Communication with VSCode Team: Reach out to the vscode-python extension team or community for insights or potential features that could aid in capturing and associating test output with specific tests within the VSCode environment.

Achieving 100% confidence in associating output with specific tests might require custom handling or modifications within the pytest framework or the VSCode extension. It's a complex task, but exploring these avenues might provide some solutions or insights.

**Reply by bluetech** [reply]
I'm assuming this is some LLM output. This is not helpful, please avoid in the future.

**Comment by RonnyPfannschmidt** [comment]
It's unclear what the ask is, the default output capture assigns the output to tests already

**Reply by eleanorjboyd** [reply]
Hi! Could you give me an example of how it assigns the output per test? I see all the output like printing and logging being printed to stdout with no correlation/ knowledge of the test that ran it.

**Reply by bluetech** [reply]
Assuming you have the `TestReport` from the hook, the captured output is available in [`report.capstderr`](https://docs.pytest.org/en/stable/reference/reference.html#pytest.TestReport.capstderr) and `report.capstdout`

**Reply by jeffwright13** [reply]
...and don't forget `report.caplog`

**Comment by jeffwright13** [comment]
Curious, what plugin you are getting your info from that is not providing you what you need?

**Comment by eleanorjboyd** [comment]
Hi everyone! Thank you for your answers and input! I think my settings were configured incorrectly based on some historic code but once I got those fixed and check out `report.capstdout` and `report.caplog` that is exactly what I'm looking for. 

Sorry to bother you all and thank you for your quick replies!

**Reply by eleanorjboyd** [reply]
if anyone would like to follow along on this feature add in vscode, the following issue will track progress: https://github.com/microsoft/vscode-python/issues/22594. Thank you again and sorry for the silly question!
